design/fencing_mitigation.md

```diff
@@ -0,0 +1,34 @@
+# Problem Statement
+
+We need to ensure that only one node can be fenced at a time and that Ceph is fully recovered (has healthy state) before any fencing is initiated. The available pattern for limiting fencing is the MachineDisruptionBudget which allows us to specify maxUnavailable. However, this won’t be sufficient to ensure that Ceph has recovered before fencing is initiated.
+
+Therefore, we will control how many nodes match the MDB by dynamically adding and removing labels as will as dynamically updating the MDB. By manipulating the MDB into a state where desiredHealthy <= currentHealthy, we can disable fencing.
+
+# What is fencing?
+
+Once [MachineHealthCheck controller](https://github.com/openshift/machine-api-operator#machine-healthcheck-controller) detects that a node is `NotReady` (or some other configured condition), it will remove the associated `Machine` which will cause the node to be deleted. The `MachineSet` controller will then replace the `Machine` via the machine-api.
+
+# Why can't we use `PodDisruptionBudget`?
+
+This only applies to rook-ceph OSD pods.
+
+Any PDB with maxUnavailable >= 2 can cause storage downtime by having more than one node with OSDs down.
+
+Any PDB with maxUnavailable < 2 will completely block a node from being drained or fenced.
+
+Fencing usually happens when the node can't be communicated with, so it might not even attempt to use the eviction api before fencing making PDB useless.
+
+# Design:
+
+## node-controller
+Each `Machine` part of a `StorageCluster` will be labeled with the key `ocs.openshift.io/<cluster-name>`, and it can have the values `connected` or `disconnected`.
+
+The node label controller will label a `Machine` `disconnected` if/when all the osds scheduled there are down. This will allow fencing to happen on those nodes.
```

Comment:
Why have a "disconnected" state and not just remove the label? Is it necessary for proper sequencing of the reconciliation?

Please add an explanation of how the operator will know which OSDs are down, and what node they are on.

Reply:
Will do.